ui/src/selection-panel/selection-panel-geometry-section.tsx

```diff
@@ -143,11 +150,33 @@ const SelectionPanelGeometrySection: React.FC<GeometryPlansPanelProps> = ({
         if (visiblePlansInView.length > 0) {
             visiblePlansInView.forEach(onTogglePlanVisibility);
         } else {
-            planHeadersInView.forEach((ph) => {
-                loadPlanLayout(ph.id).then((p) => {
-                    if (p) onTogglePlanVisibility(wholePlanVisibility(p));
+            planHeadersInView.forEach((h) => startLoadingPlan(h.id));
+            getTrackLayoutPlans(
+                planHeadersInView.map((h) => h.id),
+                changeTimes.geometryPlan,
+            )
+                .then((plans) => {
+                    getPlanLinkStatuses(
+                        plans.map((p) => p.id),
+                        publishType,
+                    ).then((statuses) => {
+                        const map = new Map(loadedPlans);
+                        plans.forEach((plan) => {
+                            const status = statuses.find((s) => s.id === plan.id);
+                            if (status) {
+                                map.set(plan.id, {
+                                    planLayout: plan,
+                                    linkStatus: status,
+                                });
+                                onTogglePlanVisibility(wholePlanVisibility(plan));
+                            }
+                        });
+                        setLoadedPlanMap(map);
+                    });
+                })
+                .finally(() => {
+                    planHeadersInView.forEach((h) => finishLoadingPlan(h.id));
```

Comment:
Tämä finally tapahtuu nähdäkseni nyt vain ulommalle promiselle, ja suoritetaan kun sisempää vasta aloitetaan suorittamaan. Jos tuo sisempi kutsu palauttaisi promisensa myös, then tulos mappautuisi promiseksi joka valmistuu vasta kun kaikki on valmista. Toinen, perinteisempi, malli olisi että ketjutetaan ulomman tason then blokkeja joissa mapataan dataa ja vasta viimeisessä thenissä tehdään tuo settaus. Tällöin finally myös tarkoittaa finally. 

Varmaan nätein tulisi jos tuosta pilkkoisi oikeasti vähän funktioita sisäkkäin latomisen sijaan, eli siis jotain tyyliin:
```
getTrackLayoutPlans(..)
  .then(plans => getPlanLinkStatuses(plans, publishType)
  .then(planStatusesMap => setLoadedPlanMap(planStatusesMap))
  .finally(...)
  
function getPlanLinkStatuses(plans: GeometryPlan[], publishType: PublishType): Promise<...> {
... 
```

Reply:
Also tuo loadedPlan tuo mieleen termit loaded question ja loaded fries, ei niinkään lataamista siinä mielessä kun tässä varmaan tarkoitetaan. Olisikohan tuolle joku parempi nimi :thinking:

Reply:
Korjailin nämä. Korvasin tuon Loaded-termin Fetchedillä, joka on ehkä(?) kuvaavampi